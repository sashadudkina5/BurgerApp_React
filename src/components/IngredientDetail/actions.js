export const SHOW_INGREDIENT_DETAILS = 'SHOW_INGREDIENT_DETAILS';
export const HIDE_INGREDIENT_DETAILS = 'HIDE_INGREDIENT_DETAILS';

export const showIngredientDetails = (selectedIngredient) => ({
    type: SHOW_INGREDIENT_DETAILS,
    payload: selectedIngredient,
});

export const hideIngredientDetails = () => ({
    type: HIDE_INGREDIENT_DETAILS,
});