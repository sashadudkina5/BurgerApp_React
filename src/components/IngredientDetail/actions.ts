import {
  SHOW_INGREDIENT_DETAILS,
  HIDE_INGREDIENT_DETAILS,
  REOPEN_INGREDIENT_DETAILS,
} from "../../redux_services/types-of-actions";

import { IIngredientCard} from "../../utils/types";

export type TIngredientDetailsActions =
  | IShowIngredientDetails
  | IHideIngredientDetails
  | IReopenIngredientDetails;

export interface IShowIngredientDetails {
  readonly type: typeof SHOW_INGREDIENT_DETAILS;
  readonly selectedIngredient: IIngredientCard;
}

export const showIngredientDetails = (
  selectedIngredient: IIngredientCard
): IShowIngredientDetails => ({
  type: SHOW_INGREDIENT_DETAILS,
  selectedIngredient,
});

export interface IHideIngredientDetails {
  readonly type: typeof HIDE_INGREDIENT_DETAILS;
}

export const hideIngredientDetails = (): IHideIngredientDetails => ({
  type: HIDE_INGREDIENT_DETAILS,
});

export interface IReopenIngredientDetails {
  readonly type: typeof REOPEN_INGREDIENT_DETAILS;
  readonly ingredient: IIngredientCard;
}

export const reopenIngredientDetails = (
  ingredient: IIngredientCard
): IReopenIngredientDetails => ({
  type: REOPEN_INGREDIENT_DETAILS,
  ingredient,
});
