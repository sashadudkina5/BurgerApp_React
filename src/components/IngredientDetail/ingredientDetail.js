import ingredientDetailStyles from "./ingredientDetail.module.css";
import { useDispatch, useSelector } from "react-redux";
import {getIngredientDetails} from "../../redux_services/selectors";
import React from 'react'

function IngredientDetail({ingredient}) {

  const selectedIngredient = useSelector(getIngredientDetails);

  if (!selectedIngredient) {
    return null; // Return null if no ingredient is selected
  }

  return (
    <div className={ingredientDetailStyles.wrapper}>
      <img
        className={ingredientDetailStyles.image}
        src={selectedIngredient.image}
        alt={selectedIngredient.name}
      />
      <p className={ingredientDetailStyles.name}>{selectedIngredient.name}</p>
      <div className={ingredientDetailStyles.details}>
        <ul className={ingredientDetailStyles.detailsList}>
          <li className={ingredientDetailStyles.detailsItem}>
            <p className="text text_type_main-default text_color_inactive">
              Калории,ккал
            </p>
            <p className="text text_type_main-default text_color_inactive">
              {selectedIngredient.calories}
            </p>
          </li>
          <li className={ingredientDetailStyles.detailsItem}>
            <p className="text text_type_main-default text_color_inactive">
              Белки, г
            </p>
            <p className="text text_type_main-default text_color_inactive">
              {selectedIngredient.proteins}
            </p>
          </li>
          <li className={ingredientDetailStyles.detailsItem}>
            <p className="text text_type_main-default text_color_inactive">
              Жиры, г
            </p>
            <p className="text text_type_main-default text_color_inactive">
              {selectedIngredient.fat}
            </p>
          </li>
          <li className={ingredientDetailStyles.detailsItem}>
            <p className="text text_type_main-default text_color_inactive">
              Углеводы, г
            </p>
            <p className="text text_type_main-default text_color_inactive">
              {selectedIngredient.carbohydrates}
            </p>
          </li>
        </ul>
      </div>
    </div>
  );
}

export default IngredientDetail;
