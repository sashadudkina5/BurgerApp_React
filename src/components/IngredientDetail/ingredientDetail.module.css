.wrapper {
    height: auto;
    text-align: center;
    padding-left: 60px;
    padding-right: 60px;
    padding-bottom: 60px;
}

.detailsList {
    width: 520px;
    display: flex;
    flex-direction: row;
    column-gap: 20px;
    list-style-type: none;
    margin: 0;
    padding: 0;
    justify-content: center;
}

.image {
    display: block;
    width: 520px;
    margin: 0 auto;
}