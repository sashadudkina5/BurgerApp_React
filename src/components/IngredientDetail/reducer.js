import {
    SHOW_INGREDIENT_DETAILS,
    HIDE_INGREDIENT_DETAILS,
    REOPEN_INGREDIENT_DETAILS
  } from "./actions";
  

  // Исходное состояние
const initialState = {
    ingredientDetails: null,
    isIngredientDetailModalOpen: false
  };
  
  export const ingredientDetailsReducer = (state = initialState, action) => {
    switch (action.type) {
  
      case SHOW_INGREDIENT_DETAILS: {
        return {
          ...state,
          ingredientDetails: action.payload,
          isIngredientDetailModalOpen: true
        };
      }
  
      case HIDE_INGREDIENT_DETAILS: {
        return { ...state, ingredientDetails: [], isIngredientDetailModalOpen: false };
      }

      case REOPEN_INGREDIENT_DETAILS: {
        return {
          ...state,
          ingredientDetails: action.payload,
          isIngredientDetailModalOpen: true
        };
      }
  
      // Реакция на прочие типы экшенов
      default:
        return state;
    }
  };
  