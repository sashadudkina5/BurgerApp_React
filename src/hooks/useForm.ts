import React from "react";

/**
 * Custom React hook that manages form state. 
 * It initializes form fields with default values and provides a method to update these values based on user input.
 * 
 * @param {Object} inputValues - Initial values for form fields with keys representing the field names and their corresponding values. 
 * Default values are for email, password, name, token, and newPassword fields.
 * 
 * @returns {Object} An object containing:
 * - values: An object holding the current state of form values.
 * - handleChange: A function to be called on input change events, updating the form values state with the input's current value.
 * - setValues: A function to manually set the form values.
 * 
 * @example
 * const { values, handleChange, setValues } = useForm({ email: '', password: '' });
 * 
 * <input
 *   name="email"
 *   value={values.email}
 *   onChange={handleChange}
 * />
 * 
 * <input
 *   name="password"
 *   value={values.password}
 *   onChange={handleChange}
 * />
 */
export function useForm(inputValues = { email: "", password: "", name: "", token: "", newPassword: "" }) {
    const [values, setValues] = React.useState(inputValues);
  
    const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
      const { value, name } = event.target;
      setValues({ ...values, [name]: value });
    };
  
    return { values, handleChange, setValues };
  }