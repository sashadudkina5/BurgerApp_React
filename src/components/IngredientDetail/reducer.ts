import {
    SHOW_INGREDIENT_DETAILS,
    HIDE_INGREDIENT_DETAILS,
    REOPEN_INGREDIENT_DETAILS
  } from "../../redux_services/types-of-actions";

import {TIngredientDetailsActions} from "./actions";
import { IIngredientCard} from "../../utils/types";
  

  // Исходное состояние

  type TIngredietnDetailsState = {
    ingredientDetails: IIngredientCard | null;
    isIngredientDetailModalOpen: boolean;
  }

export const initialState: TIngredietnDetailsState = {
    ingredientDetails: null,
    isIngredientDetailModalOpen: false
  };
  
export const ingredientDetailsReducer = (state = initialState, action: TIngredientDetailsActions):TIngredietnDetailsState => {
    switch (action.type) {
  
      case SHOW_INGREDIENT_DETAILS: {
        return {
          ...state,
          ingredientDetails: action.selectedIngredient,
          isIngredientDetailModalOpen: true
        };
      }
  
      case HIDE_INGREDIENT_DETAILS: {
        return { ...state, ingredientDetails: null, isIngredientDetailModalOpen: false };
      }

      case REOPEN_INGREDIENT_DETAILS: {
        return {
          ...state,
          ingredientDetails: action.ingredient,
          isIngredientDetailModalOpen: true
        };
      }
  
      // Реакция на прочие типы экшенов
      default:
        return state;
    }
  };
  
  export default ingredientDetailsReducer;