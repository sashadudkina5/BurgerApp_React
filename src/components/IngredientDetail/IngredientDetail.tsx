import ingredientDetailStyles from "./IngredientDetail.module.css";
import React from "react";
import { useParams, useLocation } from "react-router-dom";
import { getListOfIngredientsArray } from "../../redux_services/selectors";
import { reopenIngredientDetails } from "../IngredientDetail/actions";
import {IIngredients, IIngredientCard} from "../../utils/types";
import { useAppSelector, useAppDispatch } from "../../hooks/dispatch-selectos"

interface IIngredientDetailProps {
  selectedIngredient: IIngredientCard | null;
}

function IngredientDetail({ selectedIngredient }: IIngredientDetailProps) {
  const dispatch = useAppDispatch();
  const routeParams = useParams();
  const data: IIngredients = useAppSelector(getListOfIngredientsArray);
  const location = useLocation();

  React.useEffect(() => {
    if (location.pathname.startsWith('/ingredients/') && routeParams.id && data) {
      const filteredArray = data.filter(obj => obj._id === routeParams.id);
      if (filteredArray.length > 0) {
        dispatch(reopenIngredientDetails(filteredArray[0]));
      }
    }
  }, [location, routeParams.id, data, dispatch]);


  if (!selectedIngredient) {
    return null;
  }

  return (
    <div className={ingredientDetailStyles.wrapper}>
      <img
        className={ingredientDetailStyles.image}
        src={selectedIngredient.image}
        alt={selectedIngredient.name}
      />
      <p className={ingredientDetailStyles.name}>{selectedIngredient.name}</p>
      <div className={ingredientDetailStyles.details}>
        <ul className={ingredientDetailStyles.detailsList}>
          <li className={ingredientDetailStyles.detailsItem}>
            <p className="text text_type_main-default text_color_inactive">
              Калории,ккал
            </p>
            <p className="text text_type_main-default text_color_inactive">
              {selectedIngredient.calories}
            </p>
          </li>
          <li className={ingredientDetailStyles.detailsItem}>
            <p className="text text_type_main-default text_color_inactive">
              Белки, г
            </p>
            <p className="text text_type_main-default text_color_inactive">
              {selectedIngredient.proteins}
            </p>
          </li>
          <li className={ingredientDetailStyles.detailsItem}>
            <p className="text text_type_main-default text_color_inactive">
              Жиры, г
            </p>
            <p className="text text_type_main-default text_color_inactive">
              {selectedIngredient.fat}
            </p>
          </li>
          <li className={ingredientDetailStyles.detailsItem}>
            <p className="text text_type_main-default text_color_inactive">
              Углеводы, г
            </p>
            <p className="text text_type_main-default text_color_inactive">
              {selectedIngredient.carbohydrates}
            </p>
          </li>
        </ul>
      </div>
    </div>
  );
}

export default IngredientDetail;
