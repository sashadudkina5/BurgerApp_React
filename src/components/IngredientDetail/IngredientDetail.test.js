import React from 'react';
import { render } from '@testing-library/react';
import IngredientDetail from './IngredientDetail';
import { useParams, useLocation } from 'react-router-dom';
import { useAppSelector } from '../../hooks/dispatch-selectos';

jest.mock('react-router-dom', () => ({
  useParams: jest.fn(),
  useLocation: jest.fn(),
}));
jest.mock('../../hooks/dispatch-selectos', () => ({
  useAppSelector: jest.fn(),
  useAppDispatch: () => jest.fn()
}));

describe('IngredientDetail', () => {
  it('renders correctly and matches snapshot', () => {
    useParams.mockReturnValue({ id: 'mockedId' });
    useLocation.mockReturnValue({ pathname: '/ingredients/mockedId' });
    useAppSelector.mockReturnValue([
      {
        _id: 'mockedId',
        image: 'mockedImage.jpg',
        name: 'Mocked Ingredient',
        calories: 100,
        proteins: 10,
        fat: 5,
        carbohydrates: 20,
      },
    ]);

    const selectedIngredient = {
      _id: 'mockedId',
      image: 'mockedImage.jpg',
      name: 'Mocked Ingredient',
      calories: 100,
      proteins: 10,
      fat: 5,
      carbohydrates: 20,
    };


    const { asFragment } = render(<IngredientDetail selectedIngredient={selectedIngredient} />);


    expect(asFragment()).toMatchSnapshot();
  });
});
