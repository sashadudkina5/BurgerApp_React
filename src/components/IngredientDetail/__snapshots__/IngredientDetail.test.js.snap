// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IngredientDetail renders correctly and matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="wrapper"
  >
    <img
      alt="Mocked Ingredient"
      class="image"
      src="mockedImage.jpg"
    />
    <p
      class="name"
    >
      Mocked Ingredient
    </p>
    <div
      class="details"
    >
      <ul
        class="detailsList"
      >
        <li
          class="detailsItem"
        >
          <p
            class="text text_type_main-default text_color_inactive"
          >
            Калории,ккал
          </p>
          <p
            class="text text_type_main-default text_color_inactive"
          >
            100
          </p>
        </li>
        <li
          class="detailsItem"
        >
          <p
            class="text text_type_main-default text_color_inactive"
          >
            Белки, г
          </p>
          <p
            class="text text_type_main-default text_color_inactive"
          >
            10
          </p>
        </li>
        <li
          class="detailsItem"
        >
          <p
            class="text text_type_main-default text_color_inactive"
          >
            Жиры, г
          </p>
          <p
            class="text text_type_main-default text_color_inactive"
          >
            5
          </p>
        </li>
        <li
          class="detailsItem"
        >
          <p
            class="text text_type_main-default text_color_inactive"
          >
            Углеводы, г
          </p>
          <p
            class="text text_type_main-default text_color_inactive"
          >
            20
          </p>
        </li>
      </ul>
    </div>
  </div>
</DocumentFragment>
`;
